File: Golang/SegmentTree.go
package SegmentTree

type SegmentTree struct {
	// range
	Start int
	End   int

	//value
	Maximum int
	Lazy    int

	//Tree
	Left  *SegmentTree
	Right *SegmentTree
}

func (this *SegmentTree) Query(s, e int) int {
	if s > this.End || e < this.Start {
		return 0
	}

	if s <= this.Start && e >= this.End {
		return this.Maximum
	}

	this.normalize()
	leftQuery := this.Left.Query(s, e)
	rightQuery := this.Right.Query(s, e)

	if leftQuery > rightQuery {
		return leftQuery
	} else {
		return rightQuery
	}
}

func (this *SegmentTree) Update(s, e, val int) {
	if s > this.End || e < this.Start {
		return
	}
	if s <= this.Start && e >= this.End {
		this.Maximum += val
		this.Lazy += val
		return
	}

	this.normalize()
	this.Left.Update(s, e, val)
	this.Right.Update(s, e, val)

	if this.Left.Maximum > this.Right.Maximum {
		this.Maximum = this.Left.Maximum
	} else {
		this.Maximum = this.Right.Maximum
	}
}

func (this *SegmentTree) normalize() {
	if this.Start < this.End {
		if this.Left == nil || this.Right == nil { // The created have already == Maximum ( has been + Lazy so dont need)
			middle := this.Start + (this.End-this.Start)/2
			this.Left = &SegmentTree{Start: this.Start, End: middle, Maximum: this.Maximum}
			this.Right = &SegmentTree{Start: middle + 1, End: this.End, Maximum: this.Maximum}
		} else if this.Lazy != 0 {
			this.Left.Maximum += this.Lazy
			this.Left.Lazy += this.Lazy
			this.Right.Maximum += this.Lazy
			this.Right.Lazy += this.Lazy
		}
	}
	this.Lazy = 0
}

File: Python/SegmentTree.py
# Each node true/false

class SegmentTree:
    def __init__(self, start, end, value = False):
        self.start = start
        self.end = end
        self.value = value
        self.have_lazy = False
        self.lazy = False
        self.left = None
        self.right = None
    
    def query(self, start, end):
        if end < self.start or start > self.end:
            return True
        
        if start <= self.start and end >= self.end:
            return self.value
        
        self.down()
        result = self.left.query(start, end) and self.right.query(start, end)
        return result
    
    def update(self, start, end, value):
        if end < self.start or start > self.end:
            return
        
        if start <= self.start and end >= self.end:
            self.value = value
            self.lazy = value
            self.have_lazy = True
            return

        self.down()
        self.left.update(start, end, value)
        self.right.update(start, end, value)
        self.value = self.left.value and self.right.value

    def down(self):
        if self.start != self.end:
            if not self.left:
                mid = self.start + (self.end - self.start) // 2
                self.left = SegmentTree(self.start, mid, self.value)
                self.right = SegmentTree(mid + 1, self.end, self.value)
            else:
                if self.have_lazy:
                    self.left.lazy = self.right.lazy = self.left.value = self.right.value = self.lazy
                    self.left.have_lazy = True
                    self.right.have_lazy = True
        self.have_lazy = False
        
